include/nbl/asset/IAssetManager.h

```diff
@@ -235,30 +244,61 @@ class IAssetManager : public core::IReferenceCounted, public core::QuitSignallin
                     break;
             }
 
-            if (!asset.getContents().empty() && 
+            if (!bundle.getContents().empty() && 
                 ((levelFlags & IAssetLoader::ECF_DONT_CACHE_TOP_LEVEL) != IAssetLoader::ECF_DONT_CACHE_TOP_LEVEL) &&
                 ((levelFlags & IAssetLoader::ECF_DUPLICATE_TOP_LEVEL) != IAssetLoader::ECF_DUPLICATE_TOP_LEVEL))
             {
-                _override->insertAssetIntoCache(asset, filename, ctx, _hierarchyLevel);
+                _override->insertAssetIntoCache(bundle, filename, ctx, _hierarchyLevel);
             }
-            else if (asset.getContents().empty())
+            else if (bundle.getContents().empty())
             {
                 bool addToCache;
-                asset = _override->handleLoadFail(addToCache, file, filename, filename, ctx, _hierarchyLevel);
-                if (!asset.getContents().empty() && addToCache)
-                    _override->insertAssetIntoCache(asset, filename, ctx, _hierarchyLevel);
+                bundle = _override->handleLoadFail(addToCache, file, filename, filename, ctx, _hierarchyLevel);
+                if (!bundle.getContents().empty() && addToCache)
+                    _override->insertAssetIntoCache(bundle, filename, ctx, _hierarchyLevel);
+            }
+
+            if (bundle.getContents().empty() /*|| none of req levels is dummy*/)
+                return bundle;
+
+            if (restoreLevels)
+            {
+                auto reloadParams = _params;
+                {
+                    using flags_t = std::underlying_type_t<IAssetLoader::E_CACHING_FLAGS>;
+                    constexpr uint32_t bitdepth = sizeof(flags_t)*8u;
+                    const flags_t zeroOutMask = (~static_cast<flags_t>(0)) >> (bitdepth - 2u*(restoreLevels + _hierarchyLevel));
+                    flags_t reloadFlags = reloadParams.cacheFlags;
+                    reloadFlags &= zeroOutMask; // make sure we never pointlessy reload levels above (_restoreLevels+_hierLevel) in reload pass
+                    // set flags for levels [_hierLevel,_hierLevel+_restoreLevels) to dont look into cache and dont put into cache
+                    reloadFlags = core::bitfieldInsert<flags_t>(reloadFlags, IAssetLoader::ECF_DUPLICATE_REFERENCES, _hierarchyLevel*2u, restoreLevels*2u);
+                    reloadParams.cacheFlags = static_cast<IAssetLoader::E_CACHING_FLAGS>(reloadFlags);
+
+                    reloadParams.restoreLevels = 0u; // make sure it wont turn into infinite recursion
+                    reloadParams.reload = true; // TODO (consider): alternative to this flag: another method in override just to let user choose asset for restore
+                }
+                IAssetLoader::SAssetLoadContext ctx(params, file);
+                auto asset = _override->chooseDefaultAsset(bundle, ctx);
+                if (!asset->isAnyDependencyDummy(restoreLevels - 1u))
+                    return bundle;
```

Comment:
well in this case you need to check entire bundle (restore from the asset manager), no place for override choosing asset to check and restore